### ISSUE TITLE
Introducing request timeout

### ISSUE THREAD

**Issue by felixputera** [fastify/fastify-reply-from#68]
## 🚀 Feature Proposal

Right now, there's no request timeout option on `fastify-reply-from`. Although there's a workaround when using `undici` via `undici.timeout` option, this is not available on `http` & `http2` request.

I will create a PR for this feature, if this is accepted.

## Motivation

Having timeout on requests is essential on many cases. 

## Example

```
proxy.register(require('fastify-reply-from'), {
  base: 'http://localhost:3001/',
  requestTimeout: 30 * 1000
})
```

**Comment by Eomm** [comment]
It should already doable, did you tried?

```
proxy.register(require('fastify-reply-from'), {
  base: 'http://localhost:3001/',
  undici: {
    connections: 100,
    pipelining: 10,
    timeout: 10000
  }
})
```

**Comment by Eomm** [comment]
Could we support the [all the request options](https://nodejs.org/api/http.html#http_http_request_url_options_callback)?

We could have this setup to have those option aligned:

```
proxy.register(require('fastify-reply-from'), {
  base: 'http://localhost:3001/',
  http: {
    timeout: 1000
  },
  unidici: {
    timeout: 1000
  }
})
```

Note: not all the options can be overwrite
https://github.com/fastify/fastify-reply-from/blob/master/lib/request.js#L61

**Comment by felixputera** [comment]
@Eomm that sounds good too!

However, we'd still need to attach the `on('timeout')` callback so that the request can be aborted and error can be thrown. 

Do we simply check inspect if `timeout` is set in `http` option, and set it the callback appropriately? That seems a little magical to me. I would like to hear your opinion.

**Comment by Eomm** [comment]
> However, we'd still need to attach the on('timeout') callback so that the request can be aborted and error can be thrown.

Yes, absolutely, otherwise it won't work

> Do we simply check inspect if timeout is set in http option, and set it the callback appropriately?

Yes, I don't think it is magical, the module lets the user write less "boring" code 👍

**Comment by felixputera** [comment]
> Yes, I don't think it is magical, the module lets the user write less "boring" code 👍

Fair enough, I can definitely see your point.

However on `http2`, there's [no `timeout` option on its `request`](https://nodejs.org/api/http2.html#http2_clienthttp2session_request_headers_options). I am thinking of doing this like:

```js
proxy.register(require('fastify-reply-from'), {
  base: 'http://localhost:3001/',
  http2: {
	// some clienthttp2session.request options
    waitForTrailers: true,

    // 'augmented' options
    timeout: 1000,
  }
})
```

What do you think of it?

**Comment by Eomm** [comment]
http2's timeout should be supported through [`sessionTimeout`](https://github.com/fastify/fastify-reply-from#sessiontimeout) already

A change of this parameter would mean a braking-change, so I would avoid that for now or change it in a following PR

**Comment by mcollina** [comment]
It’s ok as long as it is enabled by default. I’m also ok in bumping a major if needed.

**Comment by felixputera** [comment]
Yeah, but `sessionTimeout` is a timeout on the session. If we got multiple requests coming in succession, the previous requests would never get timed-out.

Ideally on `http2`, both `sessionTimeout` and `requestTimeout` (or `streamTimeout` to be exact) to be configurable. 

A non-breaking change would be keeping the `sessionTimeout` on base option, mark it as deprecated, and also provide the option on `http2` option:

```
proxy.register(require('fastify-reply-from'), {
  base: 'http://localhost:3001/',
  http2: {
	// some clienthttp2session.request options
    waitForTrailers: true,

    // fastify-reply-from custom options
    requestTimeout: 60000,
    requestTimeout: 5000,
  },
  // deprecated
  requestTimeout: 60000,
})
```

As per @mcollina suggestion, I definitely think that having some defaults on these values would sensible too.

**Comment by mcollina** [comment]
I'm unsure if Node.js core is providing a per-stream timeout. Can you confirm it does?

**Comment by felixputera** [comment]
Yeah I think so: https://nodejs.org/api/http2.html#http2_http2stream_settimeout_msecs_callback

**Comment by mcollina** [comment]
+1 to the approach!

**Comment by felixputera** [comment]
Closed by #71
